src/garage/torch/algos/vpg.py

```diff
@@ -128,92 +134,168 @@ def train_once(self, itr, paths):
+            tabular.record('/LossBefore', loss_before.item())
+            tabular.record('/LossAfter', loss_after.item())
+            tabular.record('/dLoss', loss_before.item() - loss_after.item())
+            tabular.record('/KLBefore', kl_before.item())
+            tabular.record('/KL', kl_after.item())
+            tabular.record('/Entropy', policy_entropy.mean().item())
 
         self._old_policy.load_state_dict(self.policy.state_dict())
 
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self.env_spec, paths),
+            discount=self.discount)
+        return np.mean(undiscounted_returns)
+
+    def _train(self, obs, actions, rewards, advantages):
+        """Train the algorithm with minibatch.
+
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+                with shape :math:`(batch_size, observation_dim)`.
+            actions (torch.Tensor): Actions fed to the environment.
+                with shape :math:`(batch_size, action_dim)`.
+            rewards (torch.Tensor): Acquired rewards.
+                with shape :math:`(batch_size, )`.
+            advantages (torch.Tensor): Advantage value at each step.
+                with shape :math:`(batch_size, )`.
+
+        Returns:
+            torch.Tensor: Calculated mean scalar value of loss
+
+        """
+        loss = self._compute_loss_with_adv(obs, actions, rewards, advantages)
+
         self._optimizer.zero_grad()
         loss.backward()
 
-        kl_before = self._compute_kl_constraint(obs).detach()
-        self._optimize(itr, obs, actions, rewards, valids, baselines)
+        self._optimize(obs, actions, rewards, advantages)
 
-        with torch.no_grad():
-            loss_after = self._compute_loss(itr, obs, actions, rewards, valids,
-                                            baselines)
-            kl = self._compute_kl_constraint(obs)
-            policy_entropy = self._compute_policy_entropy(obs)
+        return loss
 
-        average_returns = log_performance(itr,
-                                          TrajectoryBatch.from_trajectory_list(
-                                              self.env_spec, paths),
-                                          discount=self.discount)
+    def _compute_loss(self, obs, actions, rewards, valids, baselines):
+        """Compute mean value of loss.
 
-        with tabular.prefix(self.policy.name):
-            tabular.record('LossBefore', loss.item())
-            tabular.record('LossAfter', loss_after.item())
-            tabular.record('dLoss', loss.item() - loss_after.item())
-            tabular.record('KLBefore', kl_before.item())
-            tabular.record('KL', kl.item())
-            tabular.record('Entropy', policy_entropy.mean().item())
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+                with shape :math:`(N, max_path_length, observation_dim)`.
```

Comment:
nit: A* and O*

you will need a variable for max_path_length. you can define it in the `Notes:` section of the docstring.